File: 백준/Gold/25498. 핸들 뭘로 하지/핸들 뭘로 하지.cc
#include <iostream>
#include <queue>
#include <string>
#include<algorithm>
#define endl '\n'
using namespace std;

struct box{
    char c;
    int idx;
    bool operator<(const box b) const{
        return c>b.c;
    }
};
int main(){
    ios::sync_with_stdio(false); cin.tie(NULL);
    int N;
    vector<vector<int>> v;
    string s;
    cin>>N;
    cin>>s;
    v.resize(N+1);

    for(int i = 0; i<N-1;i++){
        int a,b;
        cin>>a>>b;
        v[a].push_back(b);
        v[b].push_back(a);
    }

    vector<int> visited(N+1);
    vector<char> ans;
    ans.push_back(s[0]);

    queue<pair<int,int>> q;
    q.push({1,1});
    visited[1] = 1;
    while(!q.empty()){
        pair<int,int> now = q.front(); q.pop();
        vector<pair<int,int>> vp;
        vp.push_back(now);
        while(q.front().second == now.second){
            vp.push_back(q.front());
            q.pop();
        }

        vector<box> temp;
        for(auto a: vp){
            for(auto b: v[a.first]){
                if(visited[b]==0){
                    temp.push_back({s[b-1], b});
                    visited[b] = 1;
                }
            }
        }

        sort(temp.begin(), temp.end());
        //q.push({temp[0].c, now.second+1});
        if(temp.size()>0)
        ans.push_back(temp[0].c);
        for(auto a: temp){
            if(a.c == temp[0].c){
                q.push({a.idx,now.second+1});
            }
        }
        
    }
    for(auto a: ans){
        cout<<a;
    }
}

File: 백준/Gold/25498. 핸들 뭘로 하지/핸들 뭘로 하지.py
from collections import deque

# Class to represent a box
class Box:
    def __init__(self, c, idx):
        self.c = c
        self.idx = idx

    # Custom comparison for sorting in descending order
    def __lt__(self, other):
        return self.c > other.c

# Input
N = int(input())
s = input()
v = [[] for _ in range(N + 1)]

for _ in range(N - 1):
    a, b = map(int, input().split())
    v[a].append(b)
    v[b].append(a)

visited = [0] * (N + 1)
ans = []

# Initialize
ans.append(s[0])
q = deque([(1, 1)])
visited[1] = 1

while q:
    # Pair (node, depth)
    now = q.popleft()
    vp = [now]

    # Pop all nodes that have the same depth and store in list "vp"
    while q and q[0][1] == now[1]:
        vp.append(q.popleft())

    # Push every child node to list temp and sort
    temp = []
    for a in vp:
        for b in v[a[0]]:
            if visited[b] == 0:
                temp.append(Box(s[b - 1], b))
                visited[b] = 1

    temp.sort()

    # If there is no child, continue
    if not temp:
        continue

    # Add the biggest child to the answer
    ans.append(temp[0].c)

    # Push every node that has the same child value
    for a in temp:
        if a.c == temp[0].c:
            q.append((a.idx, now[1] + 1))

# Output
print(''.join(ans))
